packages/plugin-workspaces-tools/sources/commands/foreach.ts

```diff
@@ -0,0 +1,128 @@
+          await project.resolveEverything({lockfileOnly: true, cache, report});
+
+          let workspaces = command.toLocaleLowerCase() === `run`
+            ? project.workspaces.filter(workspace => workspace.manifest.scripts.has(args[0]))
+            : project.workspaces;
+
+          if (include.length > 0) {
+            workspaces = workspaces.filter(workspace => include.includes(workspace.locator.name))
+          }
+
+          if (exclude.length > 0) {
+            workspaces = workspaces.filter(workspace => !exclude.includes(workspace.locator.name))
+          }
+
+          for (const workspace of workspaces) {
+            const ident = structUtils.convertToIdent(workspace.locator);
+
+            needsProcessing.set(ident.identHash, workspace);
+          }
+
+          while (needsProcessing.size > 0) {
+            const commandPromises = [];
+
+            for (const [identHash, workspace] of needsProcessing) {
+              // If we are already running the command on that workspace, skip
+              if (processing.has(workspace.anchoredDescriptor.descriptorHash)) {
+                continue;
+              }
+
+              let isRunnable = true;
+
+              // By default we do topological, however we don't care of the order when running
+              // in --parallel unless also given the --with-dependencies flag
+              if (!parallel || withDependencies) {
+                for (const dep of workspace.dependencies.keys()) {
+                  if (needsProcessing.has(dep))
+                    isRunnable = false;
+                }
+              }
+
+              if (!isRunnable) {
+                continue;
+              }
+
+              processing.add(workspace.anchoredDescriptor.descriptorHash);
+
+              commandPromises.push(limit(async () => {
+                commandCount += 1;
+                await runCommand(workspace);
+                needsProcessing.delete(identHash);
+                processing.delete(workspace.anchoredDescriptor.descriptorHash);
+              }));
+            }
+
+            await Promise.all(commandPromises);
+          }
+
+          async function runCommand(workspace: Workspace) {
+            const prefix = workspaceUtils.getPrefix({ configuration, workspace, prefixed, commandCount});
+            const stdout = workspaceUtils.createStream({prefix, report, interlaced});
+            const stderr = workspaceUtils.createStream({prefix, report, interlaced});
+
+            try {
+              await clipanion.run(null, args, {
+                ...env,
+                cwd: workspace.cwd,
+                stdout: stdout.stream,
+                stderr: stderr.stream,
+              });
+            } finally {
+              stdout.stream.end();
+              stderr.stream.end();
+            }
+
+            // We wait so that report doesn't log the Finished message until its
+            // really done.
```

Comment:
It's not clear what it means to be "really done" - I suggest adding a note explaining exactly the phenomenon that occurs if we don't `await`.

Also, would it be possible instead of exporting a `promise` member in the `createStream` util to just do something like this here? The promise is a bit confusing, I think it would be better to keep it close from where it matters:

```
await Promise.all([
  new Promise(r => stdout.on(`end`, r)),
  new Promise(r => stderr.on(`end`, r)),
]);
```